### DISCUSSION TITLE
allow `URLSearchParams` in places of `ParsedUrlQueryInput`

### DISCUSSION THREAD

**Discussion by stefanprobst** [vercel/next.js#18058] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

`router.push` currently accepts an `UrlObject`, where `{ query: string | ParsedUrlQueryInput }`. it would be great if - additionally - passing in a `URLSearchParams` instance was supported. This would enable such patterns as this:

```js
<form
  onSubmit={(e) => {
    const formData = new FormData(e.currentTarget);
    router.push({ query: new URLSearchParams(formData) });
  }}
>
```

## Describe the solution you'd like

see above

## Describe alternatives you've considered

it's possible to just `.toString()` the `URLSearchParams`, but better support would be welcome

**Comment by BrendanBerkley** [comment]
Seconding this.

> It's possible to just .toString() the URLSearchParams

It is, but if you're working with a slug array, I didn't see a viable way to work with that and a stringified search param. The solve for me was to create a reducer that converted `URLSearchParams` to `ParsedUrlQueryInput`, doing the thing that the latter does where the values of duplicate keys get grouped together in an array.

It'd be a lot nicer to just pass in `URLSearchParams` and have that taken care of for me.
